### PR TITLE
Simplify hard to maintain copyright notice

### PR DESCRIPTION
**Pull request by Pierre-Sassoulas**
## Description

git is the source of truth for the copyright, copyrite (the tool) was taking exponentially longer with each release, and it's
polluting the code with sometime as much as 50 lines of names. I replaced it with a single line linking to the contributors shown by github.

Also added a pre-commit hook to check and fix the copyright notice, and fixed the existing files so they have a notice

## Type of Changes

|     | Type                   |
| --- | ---------------------- |
| ✓   | :sparkles: New feature |
| ✓   | :hammer: Refactoring   |
| ✓   | :scroll: Docs          |

### CODE REVIEW

File: tbump.toml
Comment (Pierre-Sassoulas) [inline tbump.toml:34]: This is not automated.

File: doc/release.md
Comment (Pierre-Sassoulas) [inline doc/release.md:26]: Another change that is not automated

File: .pre-commit-config.yaml
Comment (Pierre-Sassoulas) [inline .pre-commit-config.yaml:29]: Last change that is not automated.

**Review by Pierre-Sassoulas — COMMENTED** [review]
Approving without

**Comment by DanielNoord** [comment]
@Pierre-Sassoulas Could you fix the `pre-commit`?

File: astroid/brain/brain_scipy_signal.py
Comment (Pierre-Sassoulas) [inline astroid/brain/brain_scipy_signal.py:5]: @DanielNoord I can be removing the line break here. But it's hard to understand why it fail, I'm investigating. Maybe because of Windows line ending in this file. I thougth the line ending were normalized.

Comment (Pierre-Sassoulas) [reply astroid/brain/brain_scipy_signal.py:5]: (Ideally I'd like to have a line break between copyright and docstring, and I don't want to have a strange error reappear if someone add a line break later)

Comment (Pierre-Sassoulas) [reply astroid/brain/brain_scipy_signal.py:5]: Yeap, that was the reason, see https://github.com/PyCQA/astroid/pull/1443

**Comment by Pierre-Sassoulas** [comment]
Blocked because we need to merge #1443 first.

File: astroid/_ast.py
Comment (DanielNoord) [inline astroid/_ast.py:3]: Can we make the `pre-commit` to add a new line here?

Comment (Pierre-Sassoulas) [reply astroid/_ast.py:3]: It's pretty complicated, we're checking that the text exists, and if it does not then we add it. If we change the copyright.txt with a newline, it would "work" but removing a line before the docstring would make the script add everything again not replace what's existing. Imo this is "good enough" ™️ but we could modify the pre-commit hook too, what do you think ?

Comment (DanielNoord) [reply astroid/_ast.py:3]: I'm not sure what you mean "a line before the docstring". There is not docstring here right?

Comment (Pierre-Sassoulas) [reply astroid/_ast.py:3]: Say we have this file:
```python
# Copyright
from a import b
```
And the copyright.txt is this:
```shell
# Copyright


```
Then the result will be this:
```python
# Copyright

# Copyright
from a import b
```
Unless we change the way the pre-commit hook works this is not ideal.

Comment (DanielNoord) [reply astroid/_ast.py:3]: Hmm yeah, we should probably try and change the hook then. Should be fairly trivial to add a `--new-line` flag.

```python
if not token[i + 1] == token.NL:
    # add token
```

Or something 😄

Comment (Pierre-Sassoulas) [reply astroid/_ast.py:3]: It would be trivial, if the code was working with tokens like a very good docstring linter I know 😄 But it's working using bytes directly (thus the issue with line ending) https://github.com/Pierre-Sassoulas/copyright_notice_precommit/blob/main/scripts/copyright_notice.py#L71

Comment (DanielNoord) [reply astroid/_ast.py:3]: Hmm, but in your `--auto-fix` PR can't you do something like:
```python
if config.new_line:
    lines_to_write += "\n"
file.write(lines_to_write)
```

Comment (Pierre-Sassoulas) [reply astroid/_ast.py:3]: I think it's more complicated than that, you'd have to detect that some line already exists to add only a line break instead of adding everything again, you'd also have to handle the line ending for linux/windows with input copyright.txt containing line break for linux/windows (?)

Comment (DanielNoord) [reply astroid/_ast.py:3]: Hm, time for `pycopyrightformatter` I guess? 😄 

But I do think we shouldn't add a `pre-commit` hook if it doesn't really satisfy our needs. In its current form this will always require us to do manual input after `pre-commit` has been ran against a new file. Can we do this change without the `pre-commit` hook?

Comment (Pierre-Sassoulas) [reply astroid/_ast.py:3]: I think it's good enough because we have no reason to change the beginning of files and once it's done it's done. I think I can do it properly now and make sure new files are done properly later.

Comment (Pierre-Sassoulas) [reply astroid/_ast.py:3]: ie. the important thing is the copyright notice, not the line break after it

**Review by cdce8p — COMMENTED** [review]
I think it's ok to remove the list of contributors from each file. Just not sure a link to Github is all that useful. The contributor graph probably doesn't show everyone.

In pylint, we have a `CONTRIBUTORS` file. I think that would make sense a bit more sense. Tbh I don't know how much work it would be to gather all individual contributors and generate that file once. After that we could ask new ones to add their names to it themselves.

In contrast to pylint, I do however think that only name / username + email (optional) should be stored. Not necessary to mention individual contributions there.

If we do decide to go that route, we should make sure to add the filename to `license_files` in `setup.cfg` so it can be included in the distribution package.

**Comment by Pierre-Sassoulas** [comment]
> The contributor graph probably doesn't show everyone.

True. I just checked. That's a problem.

>  Tbh I don't know how much work it would be to gather all individual contributors and generate that file once.

Not much see ``git shortlog --summary --numbered --email``. We can use the copyrite config for aliases and we're good to go. I'll do that.

**Comment by Pierre-Sassoulas** [comment]
> Not much

I can't believe how much in underestimated the number of person who did 2 commits total and managed to still use a different email each time 😄

File: .pre-commit-config.yaml
Comment (DanielNoord) [inline .pre-commit-config.yaml:29]: Shouldn't this be all? Do we only add copyright to Python files?

File: astroid/nodes/const.py
Comment (DanielNoord) [inline astroid/nodes/const.py:2]: Nice 😄

File: script/.copyrigth_aliases.json
Comment (DanielNoord) [inline script/.copyrigth_aliases.json:1]: To make this more "parseable" (if for whatever reason we ever want to) shouldn't we make this a object where the `name` is the key for an object with two "sub-keys": `mails` and `authorative_mail`?

File: script/.copyrigth_aliases.json
Comment (DanielNoord) [inline script/.copyrigth_aliases.json:2]: Should this file be onder `script`? Feels a bit weird.

File: script/create_contributor_list.py
Comment (DanielNoord) [inline script/create_contributor_list.py:27]:
```suggestion
    number_of_commits: int
```

File: script/create_contributor_list.py
Comment (DanielNoord) [inline script/create_contributor_list.py:32]:
```suggestion
        """Permit sorting contributors by number of commits."""
```

File: script/create_contributor_list.py
Comment (DanielNoord) [inline script/create_contributor_list.py:33]:
```suggestion
        return self.number_of_commits.__gt__(other.number_of_commits)
```

File: script/create_contributor_list.py
Comment (DanielNoord) [inline script/create_contributor_list.py:39]:
```suggestion
            self.number_of_commits + other.number_of_commits, self.name, self.mail
```

File: script/create_contributor_list.py
Comment (DanielNoord) [inline script/create_contributor_list.py:55]:
```suggestion
        "-v", "--verbose", action="store_true", default=False, help="Display logging messages"
```

File: script/.copyrigth_aliases.json
Comment (Pierre-Sassoulas) [reply script/.copyrigth_aliases.json:2]: I moved it in the base dire like before.

File: script/.copyrigth_aliases.json
Comment (Pierre-Sassoulas) [reply script/.copyrigth_aliases.json:1]: I think it's a good idea, but I don't want to change it, ever 😄 (We'd have to change the existing copyrite alias in astroid/pylint, it's not worth it imo).

File: .pre-commit-config.yaml
Comment (Pierre-Sassoulas) [reply .pre-commit-config.yaml:29]: The issue is that if the comment (meaning ``#``) do not work the same it would be an issue. We only have copyright on the code right now.

Comment (cdce8p) [inline .pre-commit-config.yaml:29]: What do you guys think about adding it only to Python files inside `astroid/*`?
IMO there isn't much value in adding it to test files or custom scripts.

File: CONTRIBUTORS.txt
Comment (cdce8p) [inline CONTRIBUTORS.txt:6]:
```suggestion
- Claudiu Popa <pcmanticore@gmail.com>
```
I think it would make sense if `contributors-txt` would use the default git style for author information.

File: copyright.txt
Comment (cdce8p) [inline copyright.txt:3]: I'm not a huge fan of having a dedicated `copyright.txt` file besides `LICENSE` just for that.
Would it be possible to pass it as string? Or maybe simpler, we could move it into the `script` subfolder.

File: setup.py
Comment (cdce8p) [inline setup.py:3]: Case in point, this is pretty useless here.

File: tests/resources.py
Comment (cdce8p) [inline tests/resources.py:3]: I would suggest we remove the license header from all test files.

**Review by cdce8p — COMMENTED** [review]
Please add `CONTRIBUTORS.txt` to `setup.cfg` so it's included in the distribution package.
This is how pylint does it. [pylint -> setup.cfg](https://github.com/PyCQA/pylint/blob/4ca885fd8c35e6781a359cda0e4de2c0910973b6/setup.cfg#L10-L12)

File: script/.copyrigth_aliases.json
Comment (Pierre-Sassoulas) [reply script/.copyrigth_aliases.json]: Backtracking on that : the current format permit inconsistencies in the json, (see the "bot" it has two entries with 2 sets of ``mails``). I'm going to make a migration tool.

File: copyright.txt
Comment (Pierre-Sassoulas) [reply copyright.txt:3]: Let's move it to the subfolder, the pre-commit hook is using bytes which make passing a string with line break complicated.

File: tests/resources.py
Comment (Pierre-Sassoulas) [reply tests/resources.py:3]: It send the message that the test files are not copyrighted, no ? And they were originally. How about ``astroid.*`` and ``tests*`` but not custom scripts ?

Comment (cdce8p) [reply tests/resources.py:3]: I just wonder what the benefit of having it in test files would be. IMO it's more or less useless. Those files aren't part of the distribution package and mostly boilerplate. In the end, there is still the LICENSE file which applies to all files, regardless if a file has a license header or not.

_Disclaimer: I'm not a lawyer, this is not legal advice._

Comment (Pierre-Sassoulas) [reply tests/resources.py:3]: Also not a lawyer 😄 To be frank I also wondered what is the use of having it in any files at all. Apparently the need to do that was removed in march 1989 when the US joined the Berne Convention (before that anything not expressively copyrighted could be stolen !). When pylint was created the remnant of the practice of putting copyright everywhere were still very much alive, probably the reason why it's done that way. That being said, I'd like to be pretty conservative and careful here as I'd prefer to be safe than sorry in this case.  (Fun fact an actual law person in at my job told us developper to do that for internal unpublished code in 2021. Guess no lawyers ever got sacked for asking to put copyright and trademark everywhere)

File: .pre-commit-config.yaml
Comment (DanielNoord) [reply .pre-commit-config.yaml:29]: Fine with me!

File: tests/resources.py
Comment (cdce8p) [reply tests/resources.py:3]: You can't add too much usually 😄 Some projects include the whole license text in every file. Others, like Python, don't add any headers.

In our case it's much more a question what is practical. My argument would be that it's quite unlikely anyone will ever pursue the copyright or GPL license for pylint / astroid code. Much less so for test code. It's already incredibly difficult to do these things for linux, a much bigger project.

Anyway, I guess a few lines more or less don't really hurt.
_To get back to pylint. There we have added header to general test files, but not the individual functional tests. I think we should keep it that way._

Comment (Pierre-Sassoulas) [reply tests/resources.py:3]: I agree we should not touch pylint's functional files especially as it would modify all the result txt files and do a massive diff for no reason.

File: script/.copyrigth_aliases.json
Comment (Pierre-Sassoulas) [reply script/.copyrigth_aliases.json]: This is blocking the release of astroid 2.11, so it's not perfect but I'm going to do that in https://github.com/Pierre-Sassoulas/contributors-txt/issues/3... later.

Comment (DanielNoord) [reply script/.copyrigth_aliases.json]: Are you saying you're going to merge this and do a follow-up after `2.11`, or are you moving this from `2.11`?

Either way, I'd say let's move to `2.12`. Doesn't really make sense to do two commits of 100+ lines in quick succession. This is not impacting the performance of `astroid` for now.

Comment (Pierre-Sassoulas) [reply script/.copyrigth_aliases.json]: I was thinking about merging before 2.11 without changing the data structure of the json. It's not really problematic that there are duplication in the json. The fix will be a version upgrade + a migration of the json.

File: setup.py
Comment (Pierre-Sassoulas) [reply setup.py]: [62e01c1](https://github.com/PyCQA/astroid/pull/1441/commits/62e01c19ae4a90503d1f3212c062a44ac347cd35)

File: copyright.txt
Comment (Pierre-Sassoulas) [reply copyright.txt]: https://github.com/PyCQA/astroid/pull/1441/commits/c8e8991853ab217d87f3ea4ea80e25fdd8720b30

File: .pre-commit-config.yaml
Comment (Pierre-Sassoulas) [reply .pre-commit-config.yaml]: Discussed [here](https://github.com/PyCQA/astroid/pull/1441#discussion_r820155488)

**Comment by Pierre-Sassoulas** [comment]
> Please add CONTRIBUTORS.txt to setup.cfg so it's included in the distribution package.

👍 https://github.com/PyCQA/astroid/pull/1441/commits/db0bff2886dd5c176b93d208f81d64adf13e1441

File: script/.copyrigth_aliases.json
Comment (DanielNoord) [reply script/.copyrigth_aliases.json]: That does feel kind of rushed: why would we commit code/data that we know is faulty. As I said, since there is no reason to rush this in the next 24 hours and we'll likely have some more review time while we wait on the last `pylint` `2.13` PRs to be adopted/reviewed I think it's best to make this PR the _big_ copyright PR. That would also help with porting this PR to `pylint` as all open discussions would be solved.

Comment (Pierre-Sassoulas) [reply script/.copyrigth_aliases.json]: Fair enough.

Comment (Pierre-Sassoulas) [reply script/.copyrigth_aliases.json]: Done, and upgraded :)

File: CONTRIBUTORS.txt
Comment (Pierre-Sassoulas) [reply CONTRIBUTORS.txt]: Done [f77b0c3](https://github.com/PyCQA/astroid/pull/1441/commits/f77b0c3d44944ab13d80bb76c8c3d1e0bb6aa0da)

File: CONTRIBUTORS.txt
Comment (cdce8p) [inline CONTRIBUTORS.txt:13]: Should we keep `<None>` or remove it in this case?

Comment (Pierre-Sassoulas) [reply CONTRIBUTORS.txt:13]: Nice catch !

File: CONTRIBUTORS.txt
Comment (cdce8p) [inline CONTRIBUTORS.txt:12]: I was wondering about the sorting. Does it make sense to sort by number of commits? That might cause a more extensive diff.

Maybe just alphabetical would be better? Or two different sections `Team` and `Contributors`? _Section names can be improved probably._

Comment (Pierre-Sassoulas) [reply CONTRIBUTORS.txt:12]: I don't mind the little diff that will happen when Daniel surpass us all 😄 

> two different sections Team and Contributors? 

This is what is done for pylint more or less (just really out of date), but we'd need to have some configuration with the list of admin though. Is it worth it ?

Comment (cdce8p) [reply CONTRIBUTORS.txt:12]: If you're ok with the diff, then it's fine by me 🤷🏻‍♂️

Comment (Pierre-Sassoulas) [reply CONTRIBUTORS.txt:12]: Just FYI I might try to do something a little fancier for pylint to preserves the comments about what was done by whom but I did not do anything yet. I'm trying to release astroid first 😄

File: script/.contributors_aliases.json
Comment (DanielNoord) [inline script/.contributors_aliases.json:97]: Should this be normalised? It is in `CONTRIBUTORS.txt`

Comment (Pierre-Sassoulas) [reply script/.contributors_aliases.json]: [dc10b01](https://github.com/PyCQA/astroid/pull/1441/commits/dc10b0184d1565273e4c65389e7fb1b09f4bf003)

File: CONTRIBUTORS.txt
Comment (DanielNoord) [inline CONTRIBUTORS.txt:5]:
```suggestion
Maintainers
```

Comment (Pierre-Sassoulas) [reply CONTRIBUTORS.txt:5]: Urgh, costly typo. 2 commits a full release and a regeneration 😄

**Review by DanielNoord — APPROVED** [review]

File: script/.contributors_aliases.json
Comment (cdce8p) [inline script/.contributors_aliases.json:5]: A true / false value here is quite inflexible. What about the actual team / group name? That would allow for more than two groups if we ever want that.

Comment (Pierre-Sassoulas) [reply script/.contributors_aliases.json]: [c8f4f3b](https://github.com/PyCQA/astroid/pull/1441/commits/c8f4f3b5ab2004b03a0dd2a23c1626b791cc8029)

**Comment by sandrotosi** [comment]
> git is the source of truth for the copyright, 

I'm sorry i'm late to the party, and please note i'm coming from Debian's perspective, but this statement is not correct.

copyright is the list of people who can apply a license to their work (ie they will need to agree to change the license)

Now it looks like every single person in https://github.com/PyCQA/astroid/blob/main/CONTRIBUTORS.txt bears equal copyright claims to every single file in the astroid project (ie, *all* of them would need to agree to change the license of a single file, even tho they never contributed to it)

this is a worst situation than before and, while IANAL it may not even be legally appropriate.

Take my notes with caution, i just package astroid  for debian and i have no standing on how astroid is developed and maintained.

**Comment by Pierre-Sassoulas** [comment]
Thank you for the insight @sandrotosi, I'm not a lawyer either and I know this subject can be touchy ! (But a list of 20 peoples on top of file that we had to update in quadrinomial time was a lot to deal with). Would you think it's better if we specify the actual git command to get the contributor for a file for each file, and remove the misleading link to the full list of contributor ? 

I.e. a light version of that would be simply the github link to the file as in https://github.com/PyCQA/astroid/blob/main/astroid/_ast.py you have a list of 4 contributors in interface, Maybe there's an API I can find for this so we only get the actual list.

As an aside I think changing the license is never going to happen (even if I'd prefer a MIT license myself) as asking everyone is near impossible, but there's probably other use cases too ?

**Comment by sandrotosi** [comment]
sorry i didnt reply sooner

> Thank you for the insight @sandrotosi, I'm not a lawyer either and I know this subject can be touchy !

IANAL either

> (But a list of 20 peoples on top of file that we had to update in quadrinomial time was a lot to deal with).

i simply cant gauge the level of pain this caused, from an outside POV, this list could be updated just once before a release: would that help?

> Would you think it's better if we specify the actual git command to get the contributor for a file for each file, and remove the misleading link to the full list of contributor ?

i'm afraid that's not enough (the git repo is not what get's distributed, which is what matters here): the way i use to clarify copyright&license is that the copyright holders are the only legal entity that can set a given license for the piece of code they have copyright rights over.

now, if you simply state

`# Copyright (c) https://github.com/PyCQA/astroid/blob/main/CONTRIBUTORS.txt`

that has simply no legal value, because it's an ever changing file, and it does not specify a legal entity able to hold a copyright over code (a URL in this case).

But even if we would consider that file a list of contributors and ship it with the code tarball (which we do for debian, since we fetch it from the github tags, and they contain all the files in the repo), do really all the people listed in that file have legal claim to change the license of every single line of code in astroid? i think not.

Other projects i think use CLA, and/or other mechanisms to handle copyright claims, and from a personal perspective i always liked the complete list of copyright holders for each files (it makes my debian work so much easier), so i may be more than biased in this discussion.

> As an aside I think changing the license is never going to happen (even if I'd prefer a MIT license myself) as asking everyone is near impossible, but there's probably other use cases too ?

a given set of copyright holders can decided to change the license of their code, as long as it remains compatible with LGPL2.1+ (which is the global license for the whole project). It's a bit of a stretch, but legally possible

**Comment by Pierre-Sassoulas** [comment]
> that has simply no legal value, because it's an ever changing file, and it does not specify a legal entity able to hold a copyright over code (a URL in this case).

Say we specify the commit like this for each file so it's not changing anymore (https://github.com/PyCQA/astroid/blob/e5f7488f9370b8f69c3605aa1be8650c7a3d76e4/astroid/_ast.py). Would it work for you ?

Result at url:
copyrightholfder

We could maybe find an API to have a json directly.

>  CLA, and/or other mechanisms to handle copyright claims,

This could work, but for me working on the copyright does not make pylint better so I really don't want to spend time on this (unless paid).

The license has been LGPL for years, it's contaminating, so whatever is done is under LGPL. Having very detailed and exact copyright feel like law theater to be honest. And I have also a feeling it's only benefiting corporation, more precisely some over-cautious corporate lawyer that do not pay for pylint because everyone working on it do it for free anyway. I'm not going to sue some multinational company if they make a private fork and do not share the result, I won't even know about it.  This is not going to be prevented by me painfully upgrading every file with a 2 screen long list of copyright holder that everyone will then have to scroll past every time they open a file to edit.

**Pull request merged**
